core/primitives-core/src/profile.rs

```diff
@@ -1,15 +1,17 @@
 use std::fmt;
+use std::ops::{Index, IndexMut};
 
-use crate::config::{ActionCosts, ExtCosts};
-use crate::types::Gas;
+use borsh::{BorshDeserialize, BorshSerialize};
+use serde::{Deserialize, Serialize};
 
-/// Data for total gas burnt (index 0), and then each external cost and action
-type DataArray = [u64; ProfileData::LEN];
+use crate::config::{ActionCosts, ExtCosts};
 
 /// Profile of gas consumption.
-#[derive(Clone, PartialEq, Eq)]
+/// Vecs are used for forward compatible. should only append new costs and never remove old costs
+/// When add new cost, the new cost should also be append to ProfileData::ALL
+#[derive(Clone, PartialEq, Eq, BorshSerialize, BorshDeserialize)]
 pub struct ProfileData {
-    data: DataArray,
+    costs: Vec<u64>,
 }
```

Comment:
Taking https://github.com/near/nearcore/pull/4438#discussion_r688456128 into account, why do we need a Vec in the first place here? Seems `Box<[u64; LEN]>` is what's needed.

Reply:
I think the idea behind using `Vec` was to enable deserialization from arrays that are longer or shorter than the current one. Though, I think it should be implemented through custom deserialization, which would initialize the missing costs (if the array is shorter) with zeros. We should never arrive to the point when we run an older code with a newer database (it won't be compatible for tons of different reasons anyway)

Reply:
Vec has a benefit to avoid migration after adding a new cost. Vec, borsh serialized, has length encoded as first bytes. So consider this example: given the testnet database, we add a new cost in the new binary. The new cost for all old txn is 0 or more acturately, "-1" since it's not counted. In the new binary, deserialize it as vec, we can know up to which costs are available and remaining costs are considered unavailable. This reduce the need on everytime we introduce a new cost, we have to add a database migration, which is slow and unnecessary to run.

Reply:
@ailisp You don't need a database migration if you implement a custom deserialization; here is some pseudocode:

```rust
struct DataArray([u64; LEN]);

impl Deserialize for DataArray {
  fn deserialize(data) {
      let data_vec = Vec::<u64>::deserialize(data)?;
      let mut data_array = [0; LEN];
      data_array.copy_from(&data_vec[..LEN.min(data_vec.len())]);
      Ok(Self(data_array))
  }
}

struct ProfileData {
  costs: DataArray
}
```

Reply:
Sounds good then!